### ISSUE TITLE
Spotted a possible code error

### ISSUE THREAD

**Issue by souldreamer** [Sawtaytoes/OneForm#2]
Hi!

While trying to typescript-ify your code (to generate a good definition file), I noticed that you have the code:
(sorry for the different formatting, I totally used prettier with my own settings so I could work on the code more easily 😳)

```javascript
	const {
		getAllErrorMessages: getAllFieldErrorMessages,
		getErrorMessages: getFieldErrorMessages,
		setErrorMessages: setFieldErrorMessages,
		subscribeToErrorMessages: subscribeToFieldErrorMessages,
	} = useErrorMessagesState({
		errorMessages,
		onErrorMessages: ({ identifier, value }) => {
			if (value) {
				setFieldVisited(identifier);
			}
		},
		updatedErrorMessages,
	});
```

However, in `useErrorMessagesState`, you call `onErrorMessages` like this:

```javascript
			onErrorMessagesRef.current({
				errorMessages: errorMessageslist,
				identifier,
			});
```

While in pure Javascript this wouldn't error at all, you'd never have a non-falsy `value` so the `setFieldVisited` statement would never execute.

Hope this helps (and love your YT channels btw 😄)

**Comment by Sawtaytoes** [comment]
Thank you so much for digging deep and finding this bug! I've gone ahead and fixed it in `v1.7.5`.

This was definitely an oversight on my part.

Also `errorMessageslist` should be `errorMessagesList`. 🤦‍♂️

If you want to help out with anything else like documentation in GitBook or have any other suggestions, feel free to put up some PRs!

---

Glad you're enjoying the videos. Makes me want to release more 👍.

**Comment by souldreamer** [comment]
I'd love to contribute directly, but unfortunately I'm pretty much TypeScript-focused, and had to run your code through prettier with my own settings just to be able to read it comfortably 😅 (while from one of your videos I totally see how it fits your workflow, I can read it much better when it's condensed - the only time I have code side-by-side is when I'm rewriting things in the same file and need to see a reference of the old code). Not to mention separating the source files a bit (adding folders for contexts and hooks to start with)

Speaking of further contributions, should I keep opening separate issues whenever I find something, or just keep reopening this issue?

Also, what are your thoughts on adding typescript or flow to your project? Seems like adding types does catch a few errors. I'm totally willing to add you to my github repo (it's private as it's not a fork, just pretty much took the react source folder and remade it into a project, without adding the tests and storybooks for now, as I'm trying to change as little of your code as possible, just want good typings so I can use it in my own project -- I'll PR the resulting .d.ts file in any case, but having just that will sooner or later introduce mismatches between the definitions and your actual code).

Speaking of things TS caught, could you take a look at `getAllValues` in `useSymbolFunctionStore`? Given that it uses `Symbol` keys on an object, `Object.entries` seems to be the wrong approach. TS was screaming at me because the type for the first element of `Object.entries`' result is `string`, no matter the object, which led me to realize that `Object.entries` doesn't return `Symbol` keys. For that you have to use `Object.getOwnPropertySymbols`.

**Comment by souldreamer** [comment]
And another issue is:

`OneForm.jsx`
```typescript
	const { validate } = useValidationsState({
		getAllFieldNames: getRegisteredFieldNames,
		getIsReadyForValidation: getIsFieldReadyForValidation,
		getValidationType,
		getValue: getFieldValue,
		setErrorMessages: setFieldErrorMessages,
		validations,
	});
```
`getAllFieldNames` doesn't exist in the parameter object of `useValidationsState`. It's probably `getAllIdentifiers`:

`useValidationState.ts`
```typescript
const useValidationsState = ({
	getAllIdentifiers = Function.prototype,
	getIsReadyForValidation = Function.prototype,
	getValidationType = Function.prototype,
	getValue = Function.prototype,
	setErrorMessages = Function.prototype,
	validations = initialValidations,
} = {}) => { ... }
```

**Comment by Sawtaytoes** [comment]
## Issue organization
I don't mind having a bunch of open issues for each one you find. It will help keep things organized.

## TypeScript definitions file
I'd love to see a PR for TypeScript definitions. That'd be a huge help!

One of the folks I spoke to said TypeScript definitions are make-or-break for his project, so I'm aware of the issue. I wrote the library back in March when TypeScript hadn't yet become the defacto standard.

I agree, keeping a separate `.d.ts` isn't a good long-term solution, but it gets us partway there sooner than later.

## TypeScript refactor
Seeing TypeScript catch these issues, is fantastic!

We can also look into adding TypeScript to the existing project; although, I don't know the setup steps necessary to get it started with Jest, Storybook, and Rollup since I've never used it before.

We can do this refactor in a branch off the main repo, so I can still release hotfixes as necessary.

### Prettier
I'd prefer to leave out Prettier. I need to be able to read my own code, and Prettier formats in a way that makes it unreadable for me even at `printWidth: 0`.

I like auto-formatting, but not the way Prettier does it.

## `useSymbolFunctionStore`
I didn't realize I was using objects with Symbol keys. I tend to use `Map` for those. You're right, I'd need to use `Object.getOwnPropertySymbols` instead of `Object.entries`. I'm surprised the code is working as-is. Probably missing some unit tests for edge cases.

## `useValidationsState`
Whoops. Yep. Completely right. Wow...

**Comment by souldreamer** [comment]
About prettier, I didn't mean to say that you should use it, just that our coding styles are so incompatible that me making direct PRs is quite a hurdle. I can barely reason about the code as you format it, but fortunately running it through prettier worked for me.

Typescript IMHO is awesome (been using it for years now), though the typing can get pretty crazy at times. For example, here's a snippet from what I have for `useSymbolFunctionStore`: (ignore the `FunctionKey` generic for now, as I haven't quite figured out yet what the function is all about and what its type should look like)
```typescript
const initialValues: Record<symbol, Map<never, Record<string, string[]>>> = {};

const useSymbolFunctionStore = <FunctionKey = any>() => {
	type ErrorMessages = Record<string, string | string[]>;
	type FunctionsMap = Map<FunctionKey, ErrorMessages>;
	const valuesRef = useRef<Record<symbol, FunctionsMap>>(initialValues);

	const getSymbol = useCallback(
		({ groupValidation, identifierGroup }: Pick<IdentifierGroup, 'groupValidation' | 'identifierGroup'>) =>
```

However, once it's done it helps enormously, even with understanding what's really happening.

About `useSymbolFunctionStore`: looked over the most recent code, and `valuesRef.current` is still an object/dictionary. The values *are* `Map`s, but the root is an object. So `valuesRef.current.entries` will always be undefined. The `Map` is for `func` -> `errorMessages`. You still need to use `Object.getOwnPropertySymbols` to get the symbol keys for the main object/dictionary.

**Comment by Sawtaytoes** [comment]
Let's move the `useSymbolFunctionStore` problem to a separate issue, and I can address it tomorrow. I've forgotten about that particular hook, what it does, why I made it, and it looks like I forgot to document it as well 😕. I don't understand why I made it a custom hook if I wasn't gonna write unit tests for it. It was actually the cause of another `1.7.x` bug related to subforms and validations.

We can also move the TypeScript discussion to a separate issue as well.

In that TypeScript example, there are 3 things I'd need to understand:
1. Generics.
2. `Record`.
3. `Pick`.
